drivers/sensor/default_rtio_sensor.c

```diff
@@ -0,0 +1,426 @@
+ *
+ * @param[in] header The header to scan
+ * @param[in] channel The channel to search for
+ * @param[in] sample The sample index for _N_ dimensional channels
+ * @return Index of the @p channel if found or negative if not found
+ */
+static inline int check_header_contains_channel(const struct sensor_data_generic_header *header,
+						enum sensor_channel channel, int sample)
+{
+	int index = 0;
+
+	if (SENSOR_CHANNEL_3_AXIS(channel)) {
+		channel -= 3 + sample;
+	}
+
+	for (int i = 0; i < header->num_channels; ++i) {
+		if (header->channel_info[i].channel == channel) {
+			return index;
+		}
+		index += SENSOR_CHANNEL_3_AXIS(header->channel_info[i].channel) ? 3 : 1;
+	}
+	return -1;
+}
+
+/**
+ * @brief Fallback function for retrofiting old drivers to rtio
+ *
+ * @param[in] dev The sensor device to read
+ * @param[in] iodev_sqe The read submission queue event
+ */
+static void sensor_submit_fallback(const struct device *dev, struct rtio_iodev_sqe *iodev_sqe)
+{
+	const struct sensor_read_config *cfg = iodev_sqe->sqe.iodev->data;
+	const enum sensor_channel *const channels = cfg->channels;
+	const size_t num_channels = cfg->count;
+	int num_output_samples = compute_num_samples(channels, num_channels);
+	uint32_t min_buf_len = compute_min_buf_len(num_channels, num_output_samples);
+	uint64_t timestamp_ns = k_ticks_to_ns_floor64(k_uptime_ticks());
+	int rc = sensor_sample_fetch(dev);
+	uint8_t *buf;
+	uint32_t buf_len;
+
+	/* Check that the fetch succeeded */
+	if (rc != 0) {
+		LOG_ERR("Failed to fetch samples");
+		rtio_iodev_sqe_err(iodev_sqe, rc);
+		return;
+	}
+
+	/* Get the buffer for the frame, it may be allocated dynamically by the rtio context */
+	rc = rtio_sqe_rx_buf(iodev_sqe, min_buf_len, min_buf_len, &buf, &buf_len);
+	if (rc != 0) {
+		LOG_ERR("Failed to get a read buffer of size %u bytes", min_buf_len);
+		rtio_iodev_sqe_err(iodev_sqe, rc);
+		return;
+	}
+
+	/* Set the timestamp and num_channels */
+	struct sensor_data_generic_header *header = (struct sensor_data_generic_header *)buf;
+
+	header->timestamp_ns = timestamp_ns;
+	header->num_channels = num_channels;
+
+	uint8_t current_channel_index = 0;
+	q31_t *q = (q31_t *)(buf + sizeof(struct sensor_data_generic_header) +
+			     num_channels * sizeof(struct sensor_data_generic_channel));
+
+	for (size_t i = 0; i < num_channels; ++i) {
+		struct sensor_value value[3];
+		int num_samples = SENSOR_CHANNEL_3_AXIS(channels[i]) ? 3 : 1;
+
+		/* Get the current channel requested by the user */
+		rc = sensor_channel_get(dev, channels[i], value);
+		if (rc != 0) {
+			LOG_INF("Failed to get channel %d, skipping", channels[i]);
+			header->channel_info[i].channel = SENSOR_CHAN_MAX;
+			current_channel_index += num_samples;
+			continue;
+		}
+
+		/* Get the largest absolute value reading to set the scale for the channel */
+		uint32_t header_scale = 0;
+
+		for (int sample = 0; sample < num_samples; ++sample) {
+			uint32_t scale = (uint32_t)llabs((int64_t)value[sample].val1) + 1;
```

Comment:
I'm not sure I get what's the meaning of this with the casting back and forth... should there be a an inline for it with a meaningful name and a description? Think it may make it more understandable.

Reply:
Cool, should still be a helper function if you think it can be useful in other places but that should probably be a separate PR anyway.